### PR TITLE
[FIX] sale,sale_project : Avoid sending multiple email on upsell

### PR DESCRIPTION
**Pull request by robinengels**
Expected behaviour:
When you create a sale order with a project associated. When the sale order change state
from non upsell to upsell sales person should receive one email.

Current behaviour:
The sales person receive an email for every entry in the timesheet after the sale order goes
from non upsell to upsell.

Step to reproduce:
Create a sale order setup to create a project and task and the invoicing policy is ordered quantities.

Add a timesheet entry which takes it past the ordered quantity.

Create the invoice for the order.

The state of the line and order goes to upsell, and the upsell email is sent to the salesperson.

Add another line in the timesheet and the email is sent again to the salesperson

opw-2581428

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by IT-Ideas — CHANGES_REQUESTED** [review]
Hello @robinengels,

I think that you should have a look [here](https://github.com/odoo/odoo/commit/6bd8e8976608eb97fc544049a91c456253d6b092).

To me the issue is more related to this change as it seems to me that we no longer ensure that there is a change in the field value. You should have a look at the code that was there before and ensure that we only treat records where the invoice_status goes from something different from 'upselling' to 'upselling'. In order to do so, you will need to access the old value through self._origin I think.

Can you also please add a test 🙂 

Thanks 🙂

**Comment by robinengels** [comment]
@IT-Ideas can i have an r+ ?

File: addons/sale/models/sale.py [hunk 1]
```diff
@@ -511,13 +511,19 @@ def create(self, vals):
...
+        for record in self:
+            if field.name == 'invoice_status' and record._origin.invoice_status == 'upselling':
+                send_mail = False
+
         super()._compute_field_value(field)
-        if field.name != 'invoice_status' or self.env.context.get('mail_activity_automation_skip'):
+        if field.name != 'invoice_status' or self.env.context.get('mail_activity_automation_skip') or not send_mail:
             return
-
+
```
Comment (IT-Ideas) [inline addons/sale/models/sale.py:522]: Please remove this diff. It is useless 🙂

File: addons/sale/models/sale.py [hunk 2]
```diff
@@ -511,13 +511,19 @@ def create(self, vals):
...
         super()._compute_field_value(field)
-        if field.name != 'invoice_status' or self.env.context.get('mail_activity_automation_skip'):
+        if field.name != 'invoice_status' or self.env.context.get('mail_activity_automation_skip') or not send_mail:
             return
-
+        
         filtered_self = self.filtered(lambda so: so.user_id and so.invoice_status == 'upselling')
         if not filtered_self:
             return
+
```
Comment (IT-Ideas) [inline addons/sale/models/sale.py:526]: Please remove this diff. It is useless 🙂

File: addons/sale_timesheet/tests/test_sale_service.py [hunk 3]
```diff
@@ -596,4 +596,4 @@ def test_copy_billable_project_and_task(self):
 
         # copy the task
         task_copy = task.copy()
-        self.assertEqual(task_copy.sale_line_id, task.sale_line_id, "Duplicating task should keep its Sale line")
+        self.assertEqual(task_copy.sale_line_id, task.sale_line_id, "Duplicating task should keep its Sale line")
```
Comment (IT-Ideas) [inline addons/sale_timesheet/tests/test_sale_service.py:599]: It seems you have removed the empty line at the end. Can you please undo this?

File: addons/sale_timesheet/tests/__init__.py [hunk 4]
```diff
@@ -10,3 +10,4 @@
 from . import test_reporting
 from . import test_project_overview
 from . import test_project_billing_multicompany
+
```
Comment (IT-Ideas) [inline addons/sale_timesheet/tests/__init__.py:13]: It seems there is already an empty line at the end, why adding a new one?

File: addons/sale/models/sale.py [hunk 5]
```diff
@@ -511,13 +511,19 @@ def create(self, vals):
...
     def _compute_field_value(self, field):
+        send_mail = True
+        for record in self:
+            if field.name == 'invoice_status' and record._origin.invoice_status == 'upselling':
+                send_mail = False
+
         super()._compute_field_value(field)
-        if field.name != 'invoice_status' or self.env.context.get('mail_activity_automation_skip'):
+        if field.name != 'invoice_status' or self.env.context.get('mail_activity_automation_skip') or not send_mail:
             return
```
Comment (IT-Ideas) [inline addons/sale/models/sale.py:521]: So if I understand you logic, as soon as 1 of the record satisfies your condition we ignore the whole process for the whole recordset?

I think that the solution is somewhere else 🙂
I am sure you will find it!

**Review by IT-Ideas — CHANGES_REQUESTED** [review]
Hello @robinengels 
I think you should ensure that you have a correct Python linter.
There were many cases where you added or removed empty lines.
With a correct linter, you would have been warned.

Besides this, your fix is not correct in my opinion. You should check the old value (and ensure that we effectively have a value change) somewhere else... bellow 😉

File: addons/sale/models/sale.py (hunk 5 above)
Comment (robinengels) [reply addons/sale/models/sale.py:521]: I've set up the linter as suggested in the Odoo documentation but i have this kind of error in file that i never edited
`continuation line over-indented for hanging indentpycodestyle(E126)` should i also ignore them ?

Comment (IT-Ideas) [reply addons/sale/models/sale.py:521]: Yep, only apply linter changes into your code changes, it does not worth applying it were you do not edit the code.

**Comment by IT-Ideas** [comment]
@robodoo r+

File: addons/sale/models/sale.py (hunk 5 above)
Comment (robinengels) [reply addons/sale/models/sale.py:521]: I think the _origin is updated after the super().compute_field_value() so the only place where i can check if the invoice status is being modified

Comment (IT-Ideas) [reply addons/sale/models/sale.py:521]: yep you are right, then you should try to compute your subset of recordset prior to it, but surely not skip the whole process if one meets the condition 🙂

**Comment by IT-Ideas** [comment]
😨 OMG I guess that is a r+ on wrong PR that happened. I will revert this.
Sorry

**Comment by IT-Ideas** [comment]
Hello @robinengels ,
I am sorry I have r+ this PR instead of another one. Learning for me: Close reviewed PR tabs order not to r+ by mistake 🙈 .

You will have to create another branch for your bug fix.

All my apologies.

**Comment by IT-Ideas** [comment]
Once done, you can close this remote branch. I keep it in case you do not have the branch locally anymore.

**Comment by robinengels** [comment]
No problem i'll make another branch and ping you once i push it.

**Comment by IT-Ideas** [comment]
OPW further handled in #79287

**Pull request closed without merging**